### PR TITLE
clean up dask tutorial

### PR DESCRIPTION
**Pull request by philippwitte**
Clean-up of dask tutorial:

- I changed the structure of the turorial such that it closely follows the previous FWI tutorial (03)
- Models and data are directly passed to functions rather than reading them from disk via pickle
- Removed redundancies in argument passing (i.e. avoid passing redundant information)
- I tried to avoid overwriting object properties where possible. The only time where something is overwritten is in the wrapper for scipy, where I have to overwrite the current model in `geometry`. I talked to Mathias about this, but he said there is currently no setter method for `geometry` objects.

I know I changed quite a lot, but I think it makes things cleaner and easier to follow. For large-scale examples, passing the data is obviously not the way to go, but for teaching purposes I think it's okay. People can then take this as a starting point and then modify it for their means.

### CODE REVIEW

File: examples/seismic/tutorials/04_dask.ipynb [hunk 1]
```diff
@@ -483,23 +373,149 @@
...
+    "## FWI with scipy's L-BFGS\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With our parallel FWI objective function in place, we can in principle implement a wide range of gradient-based optimization algorithms for FWI, such as (stochastic) gradient descent or the nonlinear conjugate gradient method. However, many optimization algorithms, especially second order methods or algorithms for constrained optimization, are far from trivial to implement correctly from scratch. Luckily, many optimization libraries exist that we can adapt for our purposes. \n",
+    "\n",
+    "Here, we demonstrate how we can interface the scipy *optimize* package to run FWI with a limited-memory Quasi-Newton method. The scipty optimize package was not specifically designed for FWI, but this does not matter, as the library simply execpts any Python function that can be evaluated for a current model iterate `x` and returns the function value and gradient:\n",
```
Comment (navjotk) [inline examples/seismic/tutorials/04_dask.ipynb]: > scipty

typo

> library simply execpts any Python function

typo + excepts -> accepts

**Comment by navjotk** [comment]
> just to be sure: this PR supersedes #945 , correct?

No, because this still creates new `Solver` objects for each shot within an iteration - if there is a memory leak somewhere there, this version would still suffer.

File: examples/seismic/tutorials/04_dask.ipynb [hunk 2]
```diff
@@ -483,23 +373,149 @@
...
+    "    vp = model.vp.data[model.nbl:-model.nbl, model.nbl:-model.nbl]\n",
+    "    m = 1.0 / (vp.reshape(-1).astype(np.float64))**2\n",
+    "    model_error.append(np.linalg.norm((xk - m)/m))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The final preparation step before we can run our example, is the definition of box constraints for the velocity. At each iteration, the optimizer will project the current model iterate onto a feasible set of velocites as defined by the lower and upper bounds `vmin` and `vmax`. Box contraints allow us to prevent velocites from taking negative values or values that are too small or large for the stabbility criteria of our modeling stepping scheme. We define the box constraints for the velocity in $km/s$ and then convert them to squared slownesses. Furthermore, we define our initial guess `m0`:"
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply examples/seismic/tutorials/04_dask.ipynb]: a symptom that we need to run a spell checker?

**Comment by FabioLuporini** [comment]
It's OK for me to mention JUDI @philippwitte :)

**Comment by philippwitte** [comment]
@navjotk @FabioLuporini  Okay, passed all tests again and is now ready from another look from the two of you. Let me know if you need anything more!

**Comment by FabioLuporini** [comment]
LGTM. @mloubout @rhodrin any remarks?

File: examples/seismic/tutorials/04_dask.ipynb [hunk 3]
```diff
@@ -292,125 +169,103 @@
...
+   "metadata": {},
+   "source": [
+    "## Parallel Full-Waveform Inversion"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now that we know how to use Dask to implement a parallel loop around a (serial) modeling function for a single shots, we can apply the same concept to an FWI objective function, which computes the FWI function value and gradient for a given geometry and observed shot record. This function follows largely the structure in tutorial 3 and involves computing the predicted data and backpropagating the residual to compute the gradient. As we do not want to update the velocity in the area of the absorbing boundaries, we only return the gradient on the (original) physical grid."
```
Comment (georgebisbas) [inline examples/seismic/tutorials/04_dask.ipynb]: for -a- single shots

File: examples/seismic/tutorials/04_dask.ipynb [hunk 4]
```diff
@@ -483,23 +389,161 @@
...
+    "    vp = model.vp.data[model.nbl:-model.nbl, model.nbl:-model.nbl]\n",
+    "    m = 1.0 / (vp.reshape(-1).astype(np.float64))**2\n",
+    "    model_error.append(np.linalg.norm((xk - m)/m))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The final preparation step before we can run our example, is the definition of box constraints for the velocity. At each iteration, the optimizer will project the current model iterate onto a feasible set of velocites as defined by the lower and upper bounds `vmin` and `vmax`. Box contraints allow us to prevent velocites from taking negative values or values that are too small or large for the stability criteria of our modeling stepping scheme. We define the box constraints for the velocity in $km/s$ and then convert them to squared slownesses. Furthermore, we define our initial guess `m0`:"
```
Comment (georgebisbas) [inline examples/seismic/tutorials/04_dask.ipynb]: a feasible set of velocities

File: examples/seismic/tutorials/04_dask.ipynb [hunk 5]
```diff
@@ -2,221 +2,95 @@
...
+   "metadata": {},
+   "source": [
+    "## Acquisition geometry"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For the acquisition geometry, we use the same setup as in tutorial 3 and put 5 source position on one side of the model, and an array of 101 receivers on the other side. Note that now our source coordinate array (`src_coordinates`) is a 5 x 2 array, containing the shot locations of all 5 source experiments. After defining the source/receiver coordinates, we set up individual geometry objects for both the observed data (using `model`) and the predicted data (using `model0`)."
```
Comment (rhodrin) [inline examples/seismic/tutorials/04_dask.ipynb]: how about ' .. and position 5 sources on one side... '?

**Comment by rhodrin** [comment]
Looks good thanks @philippwitte! Aside from the one comment above, shouldn't some instances of 'inverse crime' be 'inversion crime'?

**Comment by philippwitte** [comment]
Thanks for the review @rhodrin !
Regarding the final comment, as far as I know the term _inverse crime_ is widely used in the seismic community, so either should be fine.

**Comment by FabioLuporini** [comment]
Thanks a lot, @philippwitte . Merged!

**Pull request merged**